packages/bot-strategy-runner/src/index.ts

```diff
@@ -0,0 +1,213 @@
+    // Else, process the log. This considers the logging level defined by the config.
+    const rejectedOutputs: any = {};
+    const validOutputs: any = {};
+    logResults.forEach((result: any) => {
+      if (result.error) rejectedOutputs[result.botIdentifier] = result;
+      else validOutputs[result.botIdentifier] = _reduceLog(result, strategyRunnerConfig);
+    });
+
+    Logger.debug({
+      at: "BotStrategyRunner",
+      message: "All strategies have finished running",
+      rejectedOutputs,
+      validOutputs
+    });
+
+    if (strategyRunnerConfig.pollingDelay === 0) {
+      Logger.debug({
+        at: "BotStrategyRunner",
+        message: "End of execution loop - terminating process"
+      });
+
+      await delay(2); // waitForLogger does not always work 100% correctly in serverless. add a delay to ensure logs are captured upstream.
+      break;
+    }
+    Logger.debug({
+      at: "BotStratergyRunner",
+      message: "End of execution loop - waiting polling delay",
+      pollingDelay: `${strategyRunnerConfig.pollingDelay} (s)`
+    });
+    await delay(Number(strategyRunnerConfig.pollingDelay));
+  }
+}
+
+// Take in the user config options including config files or config URLs.
+const processExecutionOptions = async () => {
+  const options = program
+    .option("-fc, --fileConfig <path>", "input path to JSON config file.")
+    .option("-uc, --urlConfig <path>", "url to JSON config hosted online. Private resources use access token")
+    .option("-at, --accessToken <string>", "access token to access private configs online. EG private a repo")
+    .parse(process.argv)
+    .opts();
+
+  assert(options.fileConfig != undefined || options.urlConfig != undefined, "provide a file config or a URL config");
+
+  let fileConfig: any = {},
+    urlConfig: any = {};
+
+  if (options.fileConfig) fileConfig = JSON.parse(fs.readFileSync(options.fileConfig, { encoding: "utf8" }));
+
+  if (options.urlConfig) {
+    // Fetch the config from remote URL. use the access token for authorization.
+    const response = await nodeFetch(`${options.urlConfig}`, {
+      method: "GET",
+      headers: options.accessToken
+        ? {
+            Authorization: `token ${options.accessToken}`,
+            "Content-type": "application/json",
+            Accept: "application/vnd.github.v3.raw",
+            "Accept-Charset": "utf-8"
+          }
+        : {}
+    });
+    urlConfig = await response.json(); // extract JSON from the http response
+    assert(!urlConfig.message && !urlConfig.error, `Could not fetch config! :${JSON.stringify(urlConfig)}`);
+  }
+  return mergeConfig(fileConfig, urlConfig);
+};
+
+// Returns a promise that is resolved after `seconds` delay. Used to limit how long a spoke can run for.
+const _rejectAfterDelay = (seconds: number | undefined, executionConfig: any) =>
+  new Promise((resolve, _) => {
+    setTimeout(resolve, (seconds ? seconds : defaultStrategyTimeout) * 1000, {
```

Comment:
this should be rejecting based on the function name, but i understand why you would rather resolve. we are seeing the logging system leaking into completely unrelated parts of the code.

Reply:
Agree, this is a bit confusing.

Reply:
ye, this is a great point and something I spent a few hours trying to find a better pattern on but ended up settling on this.

The main issue is if the promise rejects (or an error is thrown within the promise) then you _cant_ use bluebird promises. I tried out using node [async](https://caolan.github.io/async/v3/docs.html) and [p-limit](https://github.com/sindresorhus/p-limit) which all worked pretty well but their error handling was really poor. Node async did not have any pattern that was workable with this (that I could find). p-limit was better as you could use `promise.allSettled` which can handle errors via a `rejected` status but the code was more convoluted and even harder to read. We do something like this in the serverless hub. you can check it out here https://github.com/UMAprotocol/protocol/blob/48c290fd0600cbc35aad8c9193a31a1a692311de/packages/serverless-orchestration/src/ServerlessHub.js#L119

The pattern I ended up going for was to ensure that promises _never_ error or reject. For example, the bot-wrapper has a try-catch block wrapping any code that could error. Then, append a standard `error` prop to the response from the method and check for this after the fact. This standardizes things over all promises within this context.

My question is: how would you recommend dealing with rejected promises? what pattern would you follow if you want to handle concurrency like with a bluebird promise?

Reply:
Yeah, I hear you. Promise.allSettled not being available in bluebird is annoying and makes this inconvenient. I think your current implementation is reasonable!